### PR TITLE
Daother 8064

### PR DESCRIPTION
**Pull request by j-s-135**
## Summary of Changes:
https://rcsbpdb.atlassian.net/browse/DAOTHER-8064

- streamlined Redis transactions
- simplified config file path
- connected clients with client utils

## Screenshot of Changes (or test)
(where applicable)

### Checklist:

- [y] I followed the code style guidelines
- [y] I have commented my code, particularly in hard-to-understand areas
- [y] I have tested the code / Added unit test(s)

### CODE REVIEW

File: rcsb/app/file/uploadRequest.py [hunk 1]
```diff
@@ -90,217 +118,33 @@ async def upload(
...
-                if os.path.exists(tempPath):
-                    os.unlink(tempPath)
-                # decompress
-                if decompress:
-                    with gzip.open(filePath, "rb") as r:
-                        with open(tempPath, "wb") as w:
-                            w.write(r.read())
-                    os.replace(tempPath, filePath)
-            else:
-                raise HTTPException(status_code=500, detail="Error - missing hash")
```
Comment (piehld) [inline rcsb/app/file/uploadRequest.py:163]: Is there a reason this isn't needed anymore? How will non-resumable uploads be handled?

**Review by piehld — CHANGES_REQUESTED** [review]
Thanks for the updates @j-s-135. I definitely like the new changes to how the configuration file is handled and read in. Much cleaner than before.

I had a couple questions though about some of the code reorganization as well as a handful of functions and files getting removed. Could you please see my comments?

**Comment by piehld** [comment]
Sorry, one other thing that needs to be added is updating the version number in the [`__init__.py`](https://github.com/rcsb/py-rcsb_app_file/blob/master/rcsb/app/file/__init__.py) file and the [HISTORY.txt](https://github.com/rcsb/py-rcsb_app_file/blob/master/HISTORY.txt) file.

File: rcsb/app/client/ClientUtils.py [hunk 2]
```diff
@@ -256,17 +258,13 @@ def upload(
                     int(fileSize) - (int(mD["chunkIndex"]) * int(self.chunkSize)),
                     int(self.chunkSize),
                 )
-                tmp.truncate(packetSize)
-                tmp.seek(0)
-                tmp.write(fUpload.read(packetSize))
-                tmp.seek(0)
```
[the remark this answers is not in the document]
Comment (j-s-135) [reply rcsb/app/client/ClientUtils.py:262]: The hash check occurs at the end of the upload function in IoUtils. Presumably, it still works. I will double check, though.

File: rcsb/app/tests-client/testClientUtils.py [hunk 3]
```diff
@@ -128,66 +128,6 @@ def testSimpleUpload(self):
                     logger.exception("Failing with %s (%.4f seconds)", str(e), time.time() - startTime)
                     self.fail()
 
-    def testResumableUpload(self):
```
[the remark this answers is not in the document]
Comment (j-s-135) [reply rcsb/app/tests-client/testClientUtils.py:131]: It required the uploadChunk function, which allowed a resumable test to occur by uploading one chunk at a time. The test would upload one chunk, stop, then upload the other chunk. However, I removed the uploadChunk function. It required the regular function to be split into, not just one, but multiple functions. There was a ton of boilerplate code, making the repository unreadable. Ahsan even complained that it was hard to read and unorganized. However, without the uploadChunk function, it's hard to automate a resumable upload test, and I had to remove the progress bars from the command line client.

File: rcsb/app/tests-file/testFileDownload.py
[the remark this answers is not in the document]
Comment (j-s-135) [reply rcsb/app/tests-file/testFileDownload.py:1]: Once again, there was too much boilerplate code. The testClientUtils file already tests the download function. The repository is harder to maintain if we have to write another download test.

File: rcsb/app/tests-file/testFileUpdate.py
[the remark this answers is not in the document]
Comment (j-s-135) [reply rcsb/app/tests-file/testFileUpdate.py:1]: For the same reason.

File: rcsb/app/tests-file/testFileUpload.py
[the remark this answers is not in the document]
Comment (j-s-135) [reply rcsb/app/tests-file/testFileUpload.py:1]: Redundant.

File: rcsb/app/file/ConfigProvider.py [hunk 4]
```diff
@@ -29,9 +29,9 @@
 class ConfigProvider(SingletonClass):
     """Accessors for configuration details."""
 
-    def __init__(self, configFilePath: typing.Optional[str] = None):
+    def __init__(self):
         # ---
-        self.__configFilePath = configFilePath if configFilePath else os.environ.get("CONFIG_FILE")
```
[the remark this answers is not in the document]
Comment (j-s-135) [reply rcsb/app/file/ConfigProvider.py:34]: That's true, I will add the optional parameter.

File: rcsb/app/config/setConfig.py [hunk 5]
```diff
@@ -6,5 +6,5 @@
 
 def getConfig():
     here = os.path.abspath(os.path.dirname(__file__))
-    config_file = os.path.join(HERE, "config.yml")
+    config_file = os.path.join(here, "config.yml")
```
[the remark this answers is not in the document]
Comment (j-s-135) [reply rcsb/app/config/setConfig.py:9]: That's strange, it was fixed. Must not have added it to Git.

File: rcsb/app/file/uploadRequest.py [hunk 6]
```diff
@@ -90,217 +118,33 @@ async def upload(
...
-    if uploadId:
-        status = await ioU.getSession(uploadId)
-        if status:
-            status = str(status)
-            status = status.replace("'", '"')
-            status = json.loads(status)
-            uploadCount = status["uploadCount"]
-    else:
-        uploadId = getNewUploadId()
-    return int(uploadCount), uploadId
```
[the remark this answers is not in the document]
Comment (j-s-135) [reply rcsb/app/file/uploadRequest.py:249]: The new version does not use it.

File: rcsb/app/client/ClientUtils.py [hunk 7]
```diff
@@ -288,148 +286,7 @@ def upload(
...
-        sourceFilePath,
-        saveFilePath,
-        chunkIndex,
-        expectedChunks,
-        uploadId,
-        fullTestHash,
-        decompress,
-        allowOverwrite,
-        resumable,
-    ):
```
[the remark this answers is not in the document]
Comment (j-s-135) [reply rcsb/app/client/ClientUtils.py:372]: Explanation above.

File: rcsb/app/file/uploadRequest.py (hunk 1 above)
Comment (j-s-135) [reply rcsb/app/file/uploadRequest.py:163]: Explanation above.

File: rcsb/app/tests-file/testFileDownload.py
[the remark this answers is not in the document]
Comment (piehld) [reply rcsb/app/tests-file/testFileDownload.py:1]: I definitely understand your view on this, but part of the purpose of this repo was to serve _as_ a boiler plate. I recall a while back you had suggested creating a new repo for the purpose of the DepUI work, and perhaps that would have been a wise decision (and still might be later on, if we deem necessary) so as to keep this project in tact. 

In either case though, we definitely want to keep these test files here. Even if they add redundancy to the code, it's only in the form of tests so it doesn't clog up the main codebase, and it's generally always good to have more tests than fewer, not just for the sake of coverage but also for providing users a source of examples for how to use the classes and methods in the project.

File: HISTORY.txt [hunk 8]
```diff
@@ -16,3 +16,4 @@
                      Configuration changes to support tox 4
  1-Mar-2023  - V0.23 Updates to and reorganization of client utilities
  10-Mar-2023 - V0.24 Rewrote config file path, jwt, shell scripts, and client context
+ 18-April-2023 - V0.25 Streamlined Redis transactions
```
Comment (piehld) [inline HISTORY.txt:19]:
```suggestion
 18-Apr-2023 - V0.25 Streamlined Redis transactions
```

File: rcsb/app/tests-client/testClientUtils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (j-s-135) [reply rcsb/app/tests-client/testClientUtils.py:131]: There's only one endpoint for uploads and it receives a chunk. The client utils had two functions: one received the full file and then chunked it. That one was kept. The other received one chunk. That one was removed. Resumable uploads have a parameter, "resumable", to the aforementioned functions, which configures the upload behavior.

File: rcsb/app/tests-file/testFileDownload.py
[the remark this answers is not in the document]
Comment (j-s-135) [reply rcsb/app/tests-file/testFileDownload.py:1]: I have fixed the coverage. It was 56%. The problem was that tox.ini was badly written and nobody added the tests-client folder to the tests. I rewrote it and added the new tests-client folder. Coverage has increased to 66%. Do you still want to add redundant tests just to increase coverage?

File: rcsb/app/tests-client/testClientUtils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (piehld) [reply rcsb/app/tests-client/testClientUtils.py:131]: Thanks for explaining, that makes sense to me. Sounds like a good plan.

File: rcsb/app/tests-file/testFileDownload.py
[the remark this answers is not in the document]
Comment (piehld) [reply rcsb/app/tests-file/testFileDownload.py:1]: Good catch on the tox.ini file excluding the testClient tests. I do still think we should include these other test files though, for the other reason I explain—providing examples of package usage, at its most basic level.

```
testFileDownload.py
testFileUpdate.py
testFileUpload.py
```

File: rcsb/app/file/ConfigProvider.py [hunk 9]
```diff
@@ -29,9 +29,9 @@
 class ConfigProvider(SingletonClass):
     """Accessors for configuration details."""
 
-    def __init__(self, configFilePath: typing.Optional[str] = None):
+    def __init__(self):
```
Comment (piehld) [inline rcsb/app/file/ConfigProvider.py]:
```suggestion
    def __init__(self, configFilePath: typing.Optional[str] = None):
```

File: rcsb/app/file/ConfigProvider.py [hunk 10]
```diff
@@ -29,9 +29,9 @@
...
         # ---
-        self.__configFilePath = configFilePath if configFilePath else os.environ.get("CONFIG_FILE")
+        self.__configFilePath = os.environ.get("CONFIG_FILE")
```
Comment (piehld) [inline rcsb/app/file/ConfigProvider.py]:
```suggestion
        self.__configFilePath = configFilePath if configFilePath else os.environ.get("CONFIG_FILE")
```

**Comment by piehld** [comment]
Hi @j-s-135, I made a few minor commit suggestions. Can you accept/commit them, and confirm that you're finished making any more changes? If so, then we can merge this PR.

File: rcsb/app/file/ConfigProvider.py (hunk 9 above)
Comment (j-s-135) [reply rcsb/app/file/ConfigProvider.py]: Those changes have been made.

File: rcsb/app/file/ConfigProvider.py (hunk 10 above)
Comment (j-s-135) [reply rcsb/app/file/ConfigProvider.py]: Those changes have been made.

**Review by piehld — APPROVED** [review]
Thanks for the changes @j-s-135! Feel free to merge.

**Pull request merged**